File: bridge/sc_neurocore_engine/attention.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Drop-in replacement for

"""Drop-in replacement for sc_neurocore.layers.StochasticAttention."""

from __future__ import annotations

import numpy as np
import numpy.typing as npt

from sc_neurocore_engine.sc_neurocore_engine import StochasticAttention as _RustAttention


def _to_2d(arr: npt.ArrayLike) -> npt.NDArray[np.float64]:
    a = np.asarray(arr, dtype=np.float64)
    return a[None, :] if a.ndim == 1 else a


class StochasticAttention:
    """API-compatible with sc_neurocore.layers.StochasticAttention."""

    def __init__(self, dim_k: int, temperature: float | None = None) -> None:
        self.dim_k = dim_k
        if temperature is not None:
            self._engine = _RustAttention(dim_k, temperature)
        else:
            self._engine = _RustAttention(dim_k)

    def forward(
        self, Q: npt.ArrayLike, K: npt.ArrayLike, V: npt.ArrayLike
    ) -> npt.NDArray[np.float64]:
        return np.asarray(self._engine.forward(_to_2d(Q), _to_2d(K), _to_2d(V)), dtype=np.float64)

    def forward_softmax(
        self, Q: npt.ArrayLike, K: npt.ArrayLike, V: npt.ArrayLike
    ) -> npt.NDArray[np.float64]:
        return np.asarray(
            self._engine.forward_softmax(_to_2d(Q), _to_2d(K), _to_2d(V)), dtype=np.float64
        )

    def forward_multihead_softmax(
        self, Q: npt.ArrayLike, K: npt.ArrayLike, V: npt.ArrayLike, n_heads: int
    ) -> npt.NDArray[np.float64]:
        return np.asarray(
            self._engine.forward_multihead_softmax(_to_2d(Q), _to_2d(K), _to_2d(V), int(n_heads)),
            dtype=np.float64,
        )

    def forward_sc(
        self,
        Q: npt.ArrayLike,
        K: npt.ArrayLike,
        V: npt.ArrayLike,
        length: int = 1024,
        seed: int = 44257,
    ) -> npt.NDArray[np.float64]:
        return np.asarray(
            self._engine.forward_sc(_to_2d(Q), _to_2d(K), _to_2d(V), int(length), int(seed)),
            dtype=np.float64,
        )

    def forward_multihead(
        self, Q: npt.ArrayLike, K: npt.ArrayLike, V: npt.ArrayLike, n_heads: int
    ) -> npt.NDArray[np.float64]:
        return np.asarray(
            self._engine.forward_multihead(_to_2d(Q), _to_2d(K), _to_2d(V), int(n_heads)),
            dtype=np.float64,
        )

File: src/sc_neurocore/accel/rust/safety/attention.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for attention

#![allow(unused_variables, dead_code, non_snake_case)]

#[derive(Debug, Clone)]
pub struct StochasticAttention {
    pub dim_k: f64,
    pub temperature: f64,
}

impl StochasticAttention {
    pub fn new() -> Self {
        Self {
            dim_k: 0.0_f64,
            temperature: 1.0_f64,
        }
    }

    pub fn _ensure_2d(&self, Q: f64, K: f64, V: f64) -> f64 {
        // self,
        // Q: np.ndarray[Any, Any],
        // K: np.ndarray[Any, Any],
        // V: np.ndarray[Any, Any],
        // ) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
        // if Q.ndim == 1:
        // Q = Q[0.0, :]
        // if K.ndim == 1:
        // K = K[0.0, :]
        // if V.ndim == 1:
        // V = V[0.0, :]
        // return Q, K, V
        0.0
    }

    pub fn forward(&self, Q: f64, K: f64, V: f64) -> f64 {
        // self, Q: np.ndarray[Any, Any], K: np.ndarray[Any, Any], V: np.ndarray[
        // ) -> np.ndarray[Any, Any]:
        // Q, K, V = self._ensure_2d(Q, K, V)
        // scores = np.dot(Q, K.T)
        // row_sums = np.sum(scores, axis=1, keepdims=true)
        // row_sums[row_sums == 0] = 1.0
        // attn_weights = scores / row_sums
        // return np.dot(attn_weights, V)
        0.0
    }

    pub fn forward_softmax(&self, Q: f64, K: f64, V: f64) -> f64 {
        // self, Q: np.ndarray[Any, Any], K: np.ndarray[Any, Any], V: np.ndarray[
        // ) -> np.ndarray[Any, Any]:
        // Q, K, V = self._ensure_2d(Q, K, V)
        // scores = np.dot(Q, K.T) / self.temperature
        // scores -= scores.max(axis=1, keepdims=true)
        // exp_scores = (scores_f64).exp()
        // attn_weights = exp_scores / exp_scores.sum(axis=1, keepdims=true)
        // return np.dot(attn_weights, V)
        0.0
    }

    pub fn forward_bitstream(&self, Q: f64, K: f64, V: f64, length: f64, use_sobol: f64) -> f64 {
        // self,
        // Q: np.ndarray[Any, Any],
        // K: np.ndarray[Any, Any],
        // V: np.ndarray[Any, Any],
        // length: int = 1024,
        // use_sobol: bool = false,
        // ) -> np.ndarray[Any, Any]:
        // Q, K, V = self._ensure_2d(Q, K, V)
        // N, dk = Q.shape
        // M, dv = V.shape
        // gen = generate_sobol_bitstream if use_sobol else generate_bernoulli_bi
        // # Encode Q, K as bitstreams
        // Q_bits = np.array(
        // [[gen(float((Q[i_f64).clamp(d], 0, 1)), length) for d in range(dk)] fo
        // )  # (N, dk, L)
        0.0
    }

}

pub fn validate_attention(state: &StochasticAttention) -> bool {
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_attention_new() {
        let state = StochasticAttention::new();
        assert!(validate_attention(&state));
    }

}
